### PR TITLE
Change to non-line buffered output if output is not a TTY

### LINKED ISSUE #60673: io::Stdout should use block bufferring when appropriate
I feel like a pretty common pitfall for beginning Rust programmers is to try writing a program that uses `println!` to print a lot of lines, compare its performance to a similar program written in Python, and be (rightly) baffled at the fact that Python is substantially faster. This occurred most recently here: https://www.reddit.com/r/rust/comments/bl7j7j/hey_rustaceans_got_an_easy_question_ask_here/emx3bhm/

The reason why this happens is because `io::Stdout` unconditionally uses line buffering, regardless of whether it's being used interactively (e.g., printing to a console or a tty) or whether it's printing to a file. So if you print a lot of lines, you end up calling the `write` syscall for every line, which is quite expensive. In contrast, Python uses line buffering when printing interactively, and standard block bufferring otherwise. You can see more details on this [here](https://github.com/python/cpython/blob/6daaf3f7de78eec2c80eaa8e94e4cca54f758a30/Modules/_io/_iomodule.c#L163-L173) and [here](https://docs.python.org/3/library/sys.html#sys.stdout).

In my opinion, Rust should adopt the same policy as Python. Indeed, there is even a FIXME item for this in the code:

https://github.com/rust-lang/rust/blob/ef01f29964df207f181bd5bcf236e41372a17273/src/libstd/io/stdio.rs#L401-L404

I think this would potentially solve a fairly large stumbling block that folks run into. The CLI working group [even calls it out as a performance footgun](https://rust-lang-nursery.github.io/cli-wg/tutorial/output.html#a-note-on-printing-performance). And also [here](https://github.com/rust-lang-nursery/cli-wg/issues/29) too. Additionally, [ripgrep rolls its own handling for this](https://docs.rs/grep-cli/0.1.2/grep_cli/fn.stdout.html).

I can't think of too many appreciable downsides to doing this. It is a change in behavior. For example, if you wrote a Rust program today that printed `io::stdout()`, and the user redirected the output to a file, then the user could (for example) `tail` that output and see it updated as each line was printed. If we made `io::Stdout` use block buffering when printing to a file like this, then that behavior would change. (This is the reasoning for flags like `--line-buffered` on `grep`.)

cc @rust-lang/libs

### PR DESCRIPTION
**Pull request by tbu-**
This matches glibc behavior.

This is determined using the `isatty` function on Unixes, and not
attempted at all for other operating systems.

Fixes #60673.

### CODE REVIEW

**Comment by tbu-** [comment]
I get a 4x speed up for the worst case of the old implementation:
```rust
fn main() {
    for _ in 0..(1<<20) {
        println!("y");
    }
}
```
```
$ hyperfine './old' './new'
Benchmark #1: ./old
  Time (mean ± σ):     462.1 ms ±   4.4 ms    [User: 241.5 ms, System: 218.2 ms]
  Range (min … max):   453.9 ms … 467.4 ms    10 runs
 
Benchmark #2: ./new
  Time (mean ± σ):     109.9 ms ±   5.5 ms    [User: 107.9 ms, System: 1.3 ms]
  Range (min … max):   102.0 ms … 120.6 ms    27 runs
 
Summary
  './new' ran
    4.20 ± 0.21 times faster than './old'
```

The old program is doing 1048576 `write` syscalls, the new one 256.

Changing the program to use an infinite loop, I get a throughput of 3MB/s vs 20MB/s.

**Comment by JohnCSimon** [comment]
Ping from triage 
@tbu- checks have failed 
@KodrAus Can you please review this PR? 
Thanks.

**Comment by tbu-** [comment]
An interesting question here is what we should do `exec`. The answer is: probably nothing, like libc does.

```
#include <stdio.h>
#include <unistd.h>

int main() {
	printf("a\n");
	char *a[] = {"/bin/ls", NULL};
	execv(a[0], a);
	return 0;
}
```
This doesn't output the `a` if you pipe it into cat, because stdout will be linebuffered then.

File: src/libstd/io/stdio.rs [hunk 1]
```diff
@@ -155,6 +143,121 @@ fn handle_ebadf<T>(r: io::Result<T>, default: T) -> io::Result<T> {
...
+            StdioReader::Fake => Ok(0),
+        }
+    }
+    fn read_vectored(&mut self, bufs: &mut [io::IoSliceMut<'_>]) -> io::Result<usize> {
+        match self {
+            StdioReader::Buffered(i) => i.read_vectored(bufs),
+            StdioReader::Fake => Ok(0),
+        }
+    }
+    #[inline]
```
Comment (KodrAus) [inline src/libstd/io/stdio.rs]: Is this inline necessary?

File: src/libstd/sys/windows/stdio_uwp.rs [hunk 2]
```diff
@@ -48,6 +48,11 @@ impl Stdout {
     pub fn new() -> io::Result<Stdout> {
         Ok(Stdout)
     }
+    pub fn should_be_line_buffered(&self) -> bool {
+        // FIXME: Fill in. I don't know how to check whether output is
+        // redirected on Windows.
+        true
```
Comment (KodrAus) [inline src/libstd/sys/windows/stdio_uwp.rs]: For UWP I think we can also use `GetConsoleMode`.

File: src/libstd/sys/windows/stdio.rs [hunk 3]
```diff
@@ -246,6 +246,11 @@ impl Stdout {
     pub fn new() -> io::Result<Stdout> {
         Ok(Stdout)
     }
+    pub fn should_be_line_buffered(&self) -> bool {
+        // FIXME: Fill in. I don't know how to check whether output is
+        // redirected on Windows.
+        true
```
Comment (KodrAus) [inline src/libstd/sys/windows/stdio.rs]: On Windows there's a `GetConsoleMode` function you can call. It's [used in the `atty` library's Windows implementation](https://github.com/softprops/atty/blob/master/src/lib.rs#L51-L81). It looks like that crate also has some more complex heuristics for Windows but those might not be so important here.

File: src/libstd/io/stdio.rs (hunk 1 above)
Comment (KodrAus) [inline src/libstd/io/stdio.rs]: Is this inline necessary?

File: src/libstd/io/stdio.rs [hunk 4]
```diff
@@ -155,6 +143,121 @@ fn handle_ebadf<T>(r: io::Result<T>, default: T) -> io::Result<T> {
...
+            StdioReader::Buffered(i) => i.consume(amt),
+            StdioReader::Fake => (),
+        }
+    }
+}
+
+/// A writer that can either be line-buffered, buffered, unbuffered or fake.
+///
+/// If it is fake, all outputs just succeed and are dropped silently.
+enum StdioWriter<W: Write> {
```
Comment (KodrAus) [inline src/libstd/io/stdio.rs]: There was a comment in the original `FIXME` suggesting we should call `flush` when dropping the handle if we're panicking. Did you have any thoughts on this @tbu-?

**Review by KodrAus — COMMENTED** [review]
Thanks for working on this @tbu-!

I've left a few comments. I think we can punt on Windows for a start if you're not too familiar with the platform :)

File: src/libstd/io/stdio.rs [hunk 5]
```diff
@@ -155,6 +143,121 @@ fn handle_ebadf<T>(r: io::Result<T>, default: T) -> io::Result<T> {
...
+            StdioWriter::Unbuffered(i) => i.write(buf),
+            StdioWriter::Fake => Ok(buf.len()),
+        }
+    }
+    fn write_vectored(&mut self, bufs: &[io::IoSlice<'_>]) -> io::Result<usize> {
+        match self {
+            StdioWriter::LineBuffered(i) => i.write_vectored(bufs),
+            StdioWriter::Buffered(i) => i.write_vectored(bufs),
+            StdioWriter::Unbuffered(i) => i.write_vectored(bufs),
+            StdioWriter::Fake => Ok(bufs.iter().map(|b| b.len()).sum()),
```
Comment (bluss) [inline src/libstd/io/stdio.rs]: This sum could probably wrap around, seems like it should saturate instead (not sure what's reasonable), or return the first non-empty len.

**Comment by Dylan-DPC** [comment]
@tbu- waiting for you to answer the review

**Comment by joelpalmer** [comment]
Ping from Triage:  @tbu- any updates?

**Comment by tbu-** [comment]
Too little time right now.

**Pull request closed without merging**